python/pyarrow/tests/test_ipc.py

```diff
@@ -1153,9 +1153,17 @@ def make_batches():
     batches = iter(UserList(make_batches()))  # weakrefable
     wr = weakref.ref(batches)
 
-    with pa.ipc.RecordBatchReader.from_batches(make_schema(),
-                                               batches) as reader:
+    with pa.RecordBatchReader.from_batches(make_schema(),
+                                           batches) as reader:
         batches = None
         assert wr() is not None
         assert list(reader) == make_batches()
         assert wr() is None
+
+    # ensure we get proper error when not passing a schema
+    # (https://issues.apache.org/jira/browse/ARROW-18229)
+    batches = make_batches()
+    with pytest.raises(TypeError):
+        reader = pa.RecordBatchReader.from_batches(
+            [('field', pa.int64())], batches)
+        str(reader.schema)
```

Comment:
Why the `str` call? Which line is failing with a `TypeError`?

Reply:
With this PR, it is the `from_batches` that is already raising the TypeErrror, but the original segfault came from inspecting a null schema, so therefore I left the code that reproduced the original issue (although now it it not strictly needed)

Reply:
The problem is that the test might succeed now if `from_batches` succeeds but `str` raises. So we should remove this  `str` call, IMHO.

Reply:
> the original segfault came from inspecting a null schema

That's probably something to fix separately as well, although there is no way to create this from Python, except through a bug like this (it is the cython `shared_ptr[CSchema]()`)